src/transport/GroupPeerMessageCounter.cpp

```diff
@@ -0,0 +1,280 @@
+                    continue;
+                }
+                // Find the last non empty element
+                for (uint32_t i = CHIP_CONFIG_MAX_FABRICS - 1; i > it; i--)
+                {
+                    if (mGroupFabrics[i].mFabricIndex != kUndefinedFabricIndex)
+                    {
+                        // Logic works since all buffer are static
+                        // move it up front
+                        memcpy(static_cast<void *>(mGroupFabrics + it), static_cast<void *>(mGroupFabrics + i),
+                               sizeof(GroupFabric));
+                        // replace with empty object (easiest way to make sure everything is cleared)
+                        memcpy(static_cast<void *>(mGroupFabrics + i), static_cast<void *>(&buf), sizeof(GroupFabric));
+                        it++;
+                        break;
+                    }
+                }
+                // Nothing to move around
+                break;
+            }
+        }
+    }
+
+    // Cannot find Peer to remove
+    return err;
+}
+
+// For Unit Test
+FabricIndex GroupPeerTable::GetFabricIndexAt(uint8_t index)
+{
+    if (index < CHIP_CONFIG_MAX_FABRICS)
+    {
+        return mGroupFabrics[index].mFabricIndex;
+    }
+
+    return kUndefinedFabricIndex;
+}
+
+GroupClientCounters::GroupClientCounters(chip::PersistentStorageDelegate * storage_delegate)
+{
+    Init(storage_delegate);
+}
+
+CHIP_ERROR GroupClientCounters::Init(chip::PersistentStorageDelegate * storage_delegate)
+{
+
+    if (storage_delegate == nullptr)
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    // TODO Implement Logic for first time use / factory reset to be random
+    // Spec 4.5.1.3
+    mStorage      = storage_delegate;
+    uint16_t size = static_cast<uint16_t>(sizeof(uint32_t));
+    DefaultStorageKeyAllocator key;
+    mStorage->SyncGetKeyValue(key.GroupControlCounter(), &mGroupControlCounter, size);
+    mStorage->SyncGetKeyValue(key.GroupDataCounter(), &mGroupDataCounter, size);
+
+    uint32_t temp = mGroupControlCounter + GROUP_MSG_COUNTER_MIN_INCREMENT;
+    mStorage->SyncSetKeyValue(key.GroupControlCounter(), &temp, size);
+
+    temp = mGroupDataCounter + GROUP_MSG_COUNTER_MIN_INCREMENT;
+    mStorage->SyncSetKeyValue(key.GroupDataCounter(), &temp, size);
+
+    return CHIP_NO_ERROR;
+}
+
+uint32_t GroupClientCounters::GetCounter(bool isControl)
+{
+    return (isControl) ? mGroupControlCounter : mGroupDataCounter;
+}
+
+void GroupClientCounters::SetCounter(bool isControl, uint32_t value)
+{
+    uint32_t temp = 0;
+    uint16_t size = static_cast<uint16_t>(sizeof(uint32_t));
+    DefaultStorageKeyAllocator key;
+
+    if (mStorage == nullptr)
+    {
+        return;
+    }
+
+    if (isControl)
+    {
+        mStorage->SyncGetKeyValue(key.GroupControlCounter(), &temp, size);
+        if (temp <= value || ((temp > (UINT32_MAX - GROUP_MSG_COUNTER_MIN_INCREMENT)) && (value < GROUP_MSG_COUNTER_MIN_INCREMENT)))
```

Comment:
Done

Reply:
The problem I saw was with the usage of the `SetCounter` method. That enable the possibility of an increment greater than 1 since we trust the caller to do the right thing. I've changed the code to perform the `temp - value == 0` which makes more sense and I also replaced the `SetCounter` by `IncrementCounter` that way nothing outside the class can change the counter value with an increment greater than 1.